format-delimited/src/main/java/io/cdap/plugin/format/delimited/input/PathTrackingDelimitedInputFormat.java

```diff
@@ -30,25 +32,130 @@
+   * quote will be read as part of value. if the quotes are not at front and end of a string, for
+   * example: a, b"c,d"e and delimiter is ',', then the delimiter inside the quote won't break the
+   * value, so [a, b"c,d"e].
+   *
+   * @param delimitedString the string to split
+   * @param delimiter the separtor
+   * @return a list of splits of the original string
+   */
+  @VisibleForTesting
+  static List<String> splitQuotesString(String delimitedString, String delimiter) {
+    // Use a tree map so we can easily look up if a delimiter is within a pair of quotes. The map
+    // key is the index of the starting and the ending quote and the value is the index of the
+    // paired quote. Example:
+    // pos: 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 |
+    //      " | a | , | b | " | , | " | c | , | d | "  |
+    // key-value pairs in the map:
+    // 0 -> 4
+    // 4 -> 0
+    // 6 -> 10
+    // 10 -> 6
+    TreeMap<Integer, Integer> quotesPos = new TreeMap<>();
+    int firstQuotePos = -1;
+    for (int i = 0; i < delimitedString.length(); i++) {
+      if (delimitedString.charAt(i) != QUOTE_CHAR) {
+        continue;
+      }
+
+      if (firstQuotePos == -1) {
+        firstQuotePos = i;
+      } else {
+        quotesPos.put(i, firstQuotePos);
+        quotesPos.put(firstQuotePos, i);
+        firstQuotePos = -1;
+      }
+    }
+
+    List<String> result = new ArrayList<>();
+    StringBuilder split = new StringBuilder();
+    for (int i = 0; i < delimitedString.length(); i++) {
+      // if the length is not enough for the delimiter, just add it to split
+      if (i + delimiter.length() > delimitedString.length()) {
+        split.append(delimitedString.charAt(i));
+        continue;
+      }
+
+      // not a delimiter
+      if (!delimitedString.startsWith(delimiter, i)) {
+        split.append(delimitedString.charAt(i));
+        continue;
+      }
+
+      // find a delimiter
+      // find the prev quote and next quote index
+      Integer prevQuotePos = quotesPos.floorKey(i);
+      Integer nextQuotePos = quotesPos.ceilingKey(i);
+      // if they exist and they form a pair, we will not split, just add the current character
+      // to split
+      if (prevQuotePos != null
+          && nextQuotePos != null
+          && quotesPos.get(prevQuotePos).equals(nextQuotePos)) {
+        split.append(delimitedString, i, nextQuotePos + 1);
+        i = nextQuotePos;
+      } else {
+        // else we need to add the finding split to result, and increment the index, reset the
+        // split
+        result.add(trimQuotes(split.toString()));
+        i = i + delimiter.length() - 1;
+        split = new StringBuilder();
+      }
+    }
+
+    // add what is remaining to the result
+    result.add(trimQuotes(split.toString()));
+    return result;
+  }
+
+  /** Trim the quotes if and only if the start and end of the string are quotes */
```

Comment:
this is odd behavior to me, it means something like `a"b"c` gets read as-is but `"abc"` gets read as `abc`? Unless we have a concrete use case for this, I think it's simpler and more understandable if we always trim the quotes.

Reply:
Yeah, let's always trim the quotes